pkg/credentialprovider/plugin/plugin.go

```diff
@@ -178,6 +226,20 @@ type pluginProvider struct {
 
 	// lastCachePurge is the last time cache is cleaned for expired entries.
 	lastCachePurge time.Time
+
+	// pluginUsesServiceAccount is set true when
+	// 1. KubeletServiceAccountTokenForCredentialProviders feature gate is enabled
+	// 2. TokenAttributes is set in the credential provider config and tokenAttributes.ServiceAccountTokenAudience is set
+	pluginUsesServiceAccount bool
+
+	serviceAccount      func(namespace, name string) (*v1.ServiceAccount, error)
+	serviceAccountToken func(podName, podNamespace, serviceAccountName string, podUID types.UID) (string, error)
+
+	// serviceAccountAnnotationKeys is the list of annotation keys that the plugin is interested in.
+	// The keys defined in this list will be extracted from the corresponding service account and passed
+	// to the plugin as part of the CredentialProviderRequest. The plugin is responsible for validating
+	// the existence of annotations and their values.
+	serviceAccountAnnotationKeys []string
```

Comment:
I suspect all of these will get replaces with 'this func does the cache key' and 'this func does the actual exec' etc once you factor out the logic that is different from the two cases.